### ISSUE TITLE
Weird behaviour of Executor Queues leading to Quarantined test_impersonation_subdag  in CI 

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#28275]
### Body

I am on a quest now to get rid of the Quarantined tests in CI and I have found a case that indicates that we might have a dangerous race in the core of Airflow executors. This happens rarely and it's next to impossible to reproduce on demand, so I decided to capture it in this issue. Also GitHub Action log will be removed in the future, so I copy the logs and describe the behaviour.

The failing case is here: https://github.com/apache/airflow/actions/runs/3652820652/jobs/6171887435

This is the `tests.core.test_impersonation_tests.TestImpersonation testMethod=test_impersonation_subdag` failing - the test is about running a backfill on a subdag with impersonation and it seems that it works fine in most cases, but in some cases it looks like the same task is added twice to the executor Queue and when the task executes, the previously prepared temporary .cfg is already removed by the first execution, when the second execution attempts to start.

I looked at the code and various paths, but I could not figure how this can happen. I think it would be great to have more pairs of eyes to look at it.

Logs (I added my comments pointing to relevant lines):

```
 ----------------------------- Captured stderr call -----------------------------
  INFO  [airflow.executors.executor_loader] Loaded executor: SequentialExecutor
  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Adding to queue: ['airflow', 'tasks', 'run', 'impersonation_subdag', 'test_subdag_operation', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmp30wvhstz']
  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Adding to queue: ['airflow', 'tasks', 'run', 'impersonation_subdag.test_subdag_operation', 'exec_python_fn', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmpn88_y_ar']
  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Adding to queue: ['airflow', 'tasks', 'run', 'impersonation_subdag.test_subdag_operation', 'exec_bash_operator', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmptyfwogyc']
  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Executing command: ['airflow', 'tasks', 'run', 'impersonation_subdag', 'test_subdag_operation', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmp30wvhstz']
  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Executing command: ['airflow', 'tasks', 'run', 'impersonation_subdag.test_subdag_operation', 'exec_python_fn', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmpn88_y_ar']
  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Executing command: ['airflow', 'tasks', 'run', 'impersonation_subdag.test_subdag_operation', 'exec_bash_operator', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmptyfwogyc']
  WARNI [airflow.jobs.backfill_job.BackfillJob] Task instance <TaskInstance: impersonation_subdag.test_subdag_operation backfill__2015-01-01T00:00:00+00:00 [up_for_reschedule]> is up for reschedule
  INFO  [airflow.models.dagrun.DagRun] Marking run <DagRun impersonation_subdag.test_subdag_operation @ 2015-01-01T00:00:00+00:00: backfill__2015-01-01T00:00:00+00:00, state:running, queued_at: None. externally triggered: False> successful
  INFO  [airflow.models.dagrun.DagRun] DagRun Finished: dag_id=impersonation_subdag.test_subdag_operation, execution_date=2015-01-01T00:00:00+00:00, run_id=backfill__2015-01-01T00:00:00+00:00, run_start_date=2022-12-08 23:35:11.461535+00:00, run_end_date=2022-12-08 23:36:08.586102+00:00, run_duration=57.124567, state=success, external_trigger=False, run_type=backfill, data_interval_start=2015-01-01T00:00:00+00:00, data_interval_end=2015-01-02T00:00:00+00:00, dag_hash=None
  INFO  [airflow.jobs.backfill_job.BackfillJob] [backfill progress] | finished run 1 of 1 | tasks waiting: 1 | succeeded: 2 | running: 0 | failed: 0 | skipped: 0 | deadlocked: 0 | not ready: 0

  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Adding to queue: ['airflow', 'tasks', 'run', 'impersonation_subdag', 'test_subdag_operation', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmp4j2_55yd']
 
^^ HERE THE TASK IS ADDED TO THE QUEUE  (ONCE)

 INFO  [airflow.executors.sequential_executor.SequentialExecutor] Executing command: ['airflow', 'tasks', 'run', 'impersonation_subdag', 'test_subdag_operation', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmp4j2_55yd']

^^ HERE THE TASK IS EXECUTED FOR THE FIRST TIME

  INFO  [airflow.executors.sequential_executor.SequentialExecutor] Executing command: ['airflow', 'tasks', 'run', 'impersonation_subdag', 'test_subdag_operation', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmp4j2_55yd']

^^ HERE THE TASK IS EXECUTED FOR THE SECOND (!) TIME

vv AND HERE IT FAILS WITH EXCEPTION BECAUSE THE TEMPORARY CONFIG HAS BEEN DELETED BY THE FIRST EXECUTION

  Traceback (most recent call last):
    File "/usr/local/bin/airflow", line 33, in <module>
      sys.exit(load_entry_point('apache-airflow', 'console_scripts', 'airflow')())
    File "/opt/airflow/airflow/__main__.py", line 39, in main
      args.func(args)
    File "/opt/airflow/airflow/cli/cli_parser.py", line 52, in command
      return func(*args, **kwargs)
    File "/opt/airflow/airflow/utils/cli.py", line 108, in wrapper
      return f(*args, **kwargs)
    File "/opt/airflow/airflow/cli/commands/task_command.py", line 356, in task_run
      with open(args.cfg_path) as conf_file:
  FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmp4j2_55yd'
  ERROR [airflow.executors.sequential_executor.SequentialExecutor] Failed to execute task Command '['airflow', 'tasks', 'run', 'impersonation_subdag', 'test_subdag_operation', 'backfill__2015-01-01T00:00:00+00:00', '--local', '--pool', 'default_pool', '--subdir', 'DAGS_FOLDER/test_impersonation_subdag.py', '--cfg-path', '/tmp/tmp4j2_55yd']' returned non-zero exit status 1..
  userdel: airflow_test_user mail spool (/var/mail/airflow_test_user) not found
  ------------------------------ Captured log call -------------------------------
```

I am afraid it might be a sign of potential race in Airflow's code and I think a really close look at the problem might be useful.

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

**Comment by potiuk** [comment]
cc: @ashb @ephraimbuddy @kaxil @uranusjr Others - I think this one is worth looking at, I looked at it yesterday for a while and I  cannot see really why it could happen.

**Comment by potiuk** [comment]
Same behaviour here https://github.com/apache/airflow/actions/runs/3660108509/jobs/6187115621 - so it's not as rare.

**Comment by potiuk** [comment]
cc: @Taragolis too.

**Comment by Taragolis** [comment]
Also some kind of side effect might exeist, especially in this code
https://github.com/apache/airflow/blob/5f54009872b04c389899d9690d23e2a2c4b3cffd/tests/core/test_impersonation_tests.py#L60-L66

Changing `sys.path` not affected if module already imported but it just assuming.

Need to migrate this test to `pytest` and after that try to investigate

**Comment by Taragolis** [comment]
As option we could try to run each quarantined test in separate process by [pytest-forked](https://github.com/pytest-dev/pytest-forked)

**Comment by potiuk** [comment]
> As option we could try to run each quarantined test in separate process by [pytest-forked](https://github.com/pytest-dev/pytest-forked)

Yep. We have just a few quarantined tests and we could try it, but on the other hand it won't help us to solve the problems IF they are caused by side-effects - because we will not see them. This is why we will sometimes get back-forth when we move stuff of quarantine, because we will see it has interactions between different tests.

One more thing here - i think however this particular test is not because of side-effect but because of some internal race condition. Not sure why and how, but it feels like it, rather than side effect.

**Comment by Taragolis** [comment]
@potiuk I've just look more clear in the CI errors and found that this tests more probably killed by [pytest-timeout](https://github.com/pytest-dev/pytest-timeout/blob/7d4c413a6344edd8340628354419a324e68c6cb7/pytest_timeout.py#L402)

Other error could refers to the fact that pytest basically couldn't properly kill all Executors processes

```
    def _try_wait(self, wait_flags):
        """All callers to this function MUST hold self._waitpid_lock."""
        try:
>           (pid, sts) = os.waitpid(self.pid, wait_flags)
E           Failed: Timeout >60.0s
```

I increase timeout for this test in PR: https://github.com/apache/airflow/pull/28378#discussion_r1049451024

**Comment by potiuk** [comment]
Ah.... that's an interesting idea. So  if i understand it well, you think that the "double log" is the result of SIGALRM sent by the pytest-timeout rather than th cause of the timeout?

So far It looked to me that this "double log is really triggering the test to "hang" and timeout was just the result of it.  I can't see how timeout/SIGALRM could have caused the double "queueing" log - but maybe this is a plausible hypothesis? 

Am I right in interpreting your way of thinking?

**Comment by Taragolis** [comment]
It just hypothesis but yeah it also might be some kind of issue in SequentialExecutor and how it handle signals.

Locally I've run this test in breeze with small and no-timeout and:
1. Postgres backend with LocalExecutor
2. Postgres backend with SequentialExecutor
3. SQLite backend with SequentialExecutor

And the errors and handlings are different with Local and Sequential. Also with SequentialExecutor tests terminated after 15 seconds after timeout.

But the error with Sequential almost the same as we time to time get in the CI, [see attached logs archive](https://github.com/apache/airflow/files/10241573/logs.zip)

**Comment by potiuk** [comment]
Very useful to see those logs.

I think you might be RIGHT - it might be the pytest.timeout that actually triggers the problem not the other way round. 

I  am usually quite uneasy when I have no plausible hypothesis why things are happening.

And in this case, I do not know the exact reason, but I think we can be much closer to solving the problem and possibly even eliminate simiilar problems in the future.

I believe the problem might come from the fact that `pytest-timeout` uses SIGALARM. Or rather that we ALSO use use SIGALARM in Airflow:

This is the timeout context manager we use in Airflow:

```
class TimeoutPosix(_timeout, LoggingMixin):
    """POSIX Timeout version: To be used in a ``with`` block and timeout its content."""

    def __init__(self, seconds=1, error_message="Timeout"):
        super().__init__()
        self.seconds = seconds
        self.error_message = error_message + ", PID: " + str(os.getpid())

    def handle_timeout(self, signum, frame):
        """Logs information and raises AirflowTaskTimeout."""
        self.log.error("Process timed out, PID: %s", str(os.getpid()))
        raise AirflowTaskTimeout(self.error_message)

    def __enter__(self):
        try:
            signal.signal(signal.SIGALRM, self.handle_timeout)
            signal.setitimer(signal.ITIMER_REAL, self.seconds)
        except ValueError:
            self.log.warning("timeout can't be used in the current context", exc_info=True)

    def __exit__(self, type_, value, traceback):
        try:
            signal.setitimer(signal.ITIMER_REAL, 0)
        except ValueError:
            self.log.warning("timeout can't be used in the current context", exc_info=True)
```

And while I am not sure how (and if?) `with timeout` is uesd in the process, this could explain it - with timeout is interrupted by the SIGALRM from pytest, AirflowTaskTimeout is thrown and somethign somewhere re-runs the task and adds it to the queue (but. I am not sure what that something could be - I could not trace retrying /re-adding the task to AirflowTaskTimeout). It might also be because of managing threads/forking by Pytest (signals are handled in main thread only  even if they are received in another thread) - https://docs.python.org/3.4/library/signal.html#execution-of-python-signal-handlers


In this case I think it's plausible enought to not try to execute it fully, but just increase the timeout and see if it fixes the problem. It was appearing frequently enough to notice it quite easily in our CI, so if it gets fixed after increasing timeout, we will have a plausible theory.

**Comment by Taragolis** [comment]
Oh.  .... just realised that I've mention about [pytest-timeout](https://github.com/pytest-dev/pytest-timeout) but we use [pytest-timeout**s**](https://github.com/Scony/pytest-timeouts) 🤦‍♂️ 🤦‍♂️ 🤦‍♂️.

The main significant differences that:
- _pytest-timeout_ supports two types of [timeout methods](https://github.com/pytest-dev/pytest-timeout#timeout-methods): threads (more safe) and SIGALRM (default)
- _pytest-timeout**s**_ supports only SIGALRM

**Comment by Taragolis** [comment]
This doesn't mean that we need to switch to another library, at least _pytest-timeout**s**_ more graceful work with teardowns if compare to _pytest-timeout_ which basically could keep your environment with something related to the previous test run

**Comment by potiuk** [comment]
Closing after #28378 has been merged and #28765 is prepared to move the test out of quarantine. We can always come back here when needed.
